### PR TITLE
Integrate new symbol uploader build task.

### PR DESCRIPTION
**Pull request by mikem8361**
Symbol publishing will only happen for the pipeline builds if PB_SymbolServerPath and
PB_SymbolServerPAT are defined. PB_SymbolExpirationInDays is how long the symbol server
will keep. Default is 30 days. -1 means forever.

### CODE REVIEW

**Comment by mikem8361** [comment]
The buildtools PR https://github.com/dotnet/buildtools/pull/1821 is needed by my changes. It adds the build task package to the optional tooling list.

File: publish/publish-symbols.proj [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <Import Project="$(PackagesDir)/$(PublishSymbolsPackage.ToLower())/$(PublishSymbolsPackageVersion)/build/PublishSymbols.targets" Condition="'$(SymbolServerPath)'!=''" />
+
+  <Target Name="Build"
```
Comment (weshaggard) [inline publish/publish-symbols.proj]: Were can I have a look at this PublishSymbols target?

File: publish/publish-type.proj [hunk 2]
```diff
@@ -72,32 +72,17 @@
     </ItemGroup>
   </Target>
 
-  <Target Name="FindDownloadedPackagesForSigning">
+  <Target Name="PublishSymbols"
```
Comment (weshaggard) [inline publish/publish-type.proj]: We don't want to actually publish the symbols publicly at this point of the publish cycle we only want to push them to the blob feed. We will do a final publish of the full stack build where we will then publish symbols.

File: publish/publish-type.proj [hunk 3]
```diff
@@ -9,7 +9,7 @@
     <BuildDependsOn>
       ValidateProperties;
       DownloadFilesFromContainer;
-      SignSymbolPackages
+      PublishSymbols
```
Comment (weshaggard) [inline publish/publish-type.proj]: Does the new target actually handle signing the symbols correctly now or is that not needed any longer?

File: publish/publish.proj [hunk 4]
```diff
@@ -215,32 +215,17 @@
...
-    <ItemGroup>
-      <EmbedIndexArgPairs Include="@(SymbolPackagesToEmbedIndex -> '%(Identity) $(PublishDirectory)%(Filename)%(Extension)')" />
-    </ItemGroup>
-
-    <!-- Inline the EmbedIndex.ps1 script used in CoreFX and CoreCLR: directly call EmbedIndex. -->
-    <Exec Command="$(DotNetToolDir)\dotnet $(EmbedIndexToolDir)\tools\EmbedIndex.dll %(EmbedIndexArgPairs.Identity)" />
+    <PropertyGroup>
+      <_SymbolPackagesToPublish>@(SymbolPackagesToPublish)</_SymbolPackagesToPublish>
+    </PropertyGroup>
+    <MSBuild Projects="$(MSBuildThisFileDirectory)\publish-symbols.proj"
```
Comment (weshaggard) [inline publish/publish.proj]: Instead of doing this I would prefer we split out the symbol publishing step by it self in the publish build definition leg.

File: config.json [hunk 5]
```diff
@@ -78,7 +78,7 @@
       "values": [],
       "defaultValue": ""
     },
-    "BatchRestorePackages": {
+    "Sync": {
```
Comment (weshaggard) [inline config.json]: This seems unrelated.

File: buildpipeline/Core-Setup-Publish.json [hunk 6]
```diff
@@ -488,8 +488,25 @@
     "PB_ToolsRoot": {
       "value": "$(Build.SourcesDirectory)\\tools"
     },
-    "SymbolCatalogCertificateId": {
-      "value": "400"
+    "PB_OptionalToolSource": {
+      "value": "https://devdiv.pkgs.visualstudio.com/_packaging/dotnet-core-internal-tooling/nuget/v3/index.json "
```
Comment (weshaggard) [inline buildpipeline/Core-Setup-Publish.json]: @dagood don't we already handle the optional tools download? If not how do we currently get the embed index tool?

File: dependencies.props [hunk 7]
```diff
@@ -53,6 +53,12 @@
     <FeedTasksPackageVersion>2.1.0-prerelease-02221-02</FeedTasksPackageVersion>
   </PropertyGroup>
 
+  <!-- Publish symbol build task package -->
+  <PropertyGroup>
+    <PublishSymbolsPackage>Microsoft.SymbolUploader.Build.Task</PublishSymbolsPackage>
+    <PublishSymbolsPackageVersion>1.0.0-alpha-00001</PublishSymbolsPackageVersion>
```
Comment (weshaggard) [inline dependencies.props]: This is too tightly coupled to the version number in the buildtools repo. We need a better way.

File: buildpipeline/Core-Setup-Publish.json (hunk 6 above)
Comment (dagood) [reply buildpipeline/Core-Setup-Publish.json]: Right now it's only in `Core-Setup-Signing-Validation.json`, I believe the changes in this file are just copying it over. (Extra space is also there.)

**Comment by mikem8361** [comment]
Link to that file:

https://devdiv.visualstudio.com/DevDiv/_git/DotNet-SymStore?path=%2Fsrc%2FMicrosoft.SymbolUploader.Build.Task%2FPackageFiles%2FPublishSymbols.targets&version=GBmikem%2Fuploader&_a=contents

**Comment by mikem8361** [comment]
There is no signing involved or needed anymore.  We will be uploading the individual files in the symbol packages to symweb (via the VSTS symbol server https://microsoft.artifacts.visualstudio.com/defaultcollection/_apis/Symbol/symsrv) and eventually the public facing msdl server too.  EmbedIndex, etc. is not needed anymore.

This means we are switching from the myget symbol server to Microsoft’s.

**Comment by mikem8361** [comment]
It related because it allows sync.cmd to also restore the optional tooling packages like corefx does.

**Comment by mikem8361** [comment]
>>>>@dagood don't we already handle the optional tools download? If not how do we currently get the embed index tool?

As far as I know (Davis can confirm) in core-setup only the sign verification step uses the optional tooling (see buildpipeline\core-setup-signing-validation.json).

**Comment by mikem8361** [comment]
>>>This is too tightly coupled to the version number in the buildtools repo. We need a better way.

I used the same way as the FeedTaskPackage (see the same file). This version is the build task package from the DotNet-SymStore repo. It is unfortunate that my only change to buildtools is to add this package to the optional tooling but that what @dagood wanted instead of override optional.json like in corefx.

**Comment by mikem8361** [comment]
>>>> We don't want to actually publish the symbols publicly at this point of the publish cycle we only want to push them to the blob feed. We will do a final publish of the full stack build where we will then publish symbols.

Where (what target file or target I guess) is the final publish of the full stack build?  I was trying to put this where symbols were published using the old Symbols.target; I may have missed the mark.

I also needed the symbol nuget packages so I put the publish symbols target after they are downloaded and the DownloadDirectory property was set.

I put the PublishSymbols target in publish.proj so the optional tooling packages could be restored before trying to import its target file (that happens in publish-symbols.proj).

I like the idea of have the symbol publish as independent as possible but I didn’t what to duplicate how the symbol packages are downloaded from the feed.

**Comment by mikem8361** [comment]
@weshaggard, @dagood I've redone the changes not to use the optional tooling feed which has simplified things greatly.  I've deleted the buildtools PR and pushed my uploader build task package to the myget buildtools feed.

File: buildpipeline/Core-Setup-Publish.json [hunk 8]
```diff
@@ -488,8 +488,15 @@
...
-    "SymbolCatalogCertificateId": {
-      "value": "400"
+    "PB_SymbolServerPath": {
+      "value": "https://mikemvsts.artifacts.visualstudio.com/DefaultCollection"
+    },
+    "PB_SymbolServerPAT": {
+      "value": null,
+      "isSecret": true
+    }
+    "PB_SymbolExpirationInDays": {
```
Comment (dagood) [inline buildpipeline/Core-Setup-Publish.json]: Previous property is missing the `,`, causing this parse error.

File: publish/publish-type.proj [hunk 9]
```diff
@@ -118,4 +89,5 @@
 
   <Import Project="$(ToolsDir)VersionTools.targets" Condition="Exists('$(ToolsDir)VersionTools.targets')" />
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.targets))\dir.targets" />
-</Project>
\ No newline at end of file
+</Project>
+
```
Comment (dagood) [inline publish/publish-type.proj]: \<nit> One too many newlines.

File: dependencies.props (hunk 7 above)
Comment (eerhardt) [reply dependencies.props]: And we also need this version to be auto-updated.

Comment (weshaggard) [reply dependencies.props]: @mikem8361 correct me if I'm wrong but this package is manually created so there isn't any auto-updates for it.

Comment (eerhardt) [reply dependencies.props]: Isn't that even worse?  Where does this package live?

**Comment by mikem8361** [comment]
The package is created from the private VSTS DotNet-SymStore repo. It is manual. It is similar to the Microsoft.DotNet.Build.Tasks.Feed package (see dependency.props.

Since it is still pretty “alpha” I think I’d rather explicitly update the version for now. Once it stabilizes then I can look into how to make it autoupdate.

mikem

**Comment by eerhardt** [comment]
Since this is a closed source package, are we going to need a way to turn it "off" for https://github.com/dotnet/source-build/?  

/cc @dotnet/source-build-contrib

**Comment by dagood** [comment]
Yeah, I think we have the same problem with the Feed package in `init-tools.msbuild`, too, due to its dependencies. I'd classify both as selective BuildTools restore, filed https://github.com/dotnet/buildtools/issues/1824.
